core/src/main/java/hudson/slaves/WorkspaceList.java

```diff
@@ -298,12 +302,26 @@ public void release() {
      * {@link FilePath#createTempFile} may be used for this purpose if desired.
      * <p>The resulting directory may not exist; you may call {@link FilePath#mkdirs} on it if you need it to.
      * It may be deleted alongside the workspace itself during cleanup actions.
+     *
+     * The method may operate differently when the path points to a root directory or to a disk root in Windows.
+     * In such cases API users are recommended to allocate a temporary directory on their own in another location,
+     * e.g. by using {@link FilePath#createTempDir(String, String)}.
+     * This logic is not embedded into this method, because a remote operation on the agent is required to allocate the directory.
+     *
      * @param ws a directory such as a build workspace
-     * @return a sibling directory, for example {@code …/something@tmp} for {@code …/something}
+     * @return A sibling directory, for example {@code …/something@tmp} for {@code …/something}.
+     *         It may return {@code null} if it is not possible to return a sibling.
      * @since 1.652
      */
-    public static FilePath tempDir(FilePath ws) {
-        return ws.sibling(ws.getName() + COMBINATOR + "tmp");
+    @CheckForNull
```

Comment:
This is going to be incompatible for some callers from plugins I am afraid.

Reply:
@jglick would it be better to use `@NotNull` or not change this at all? Asking so I can better understand

Reply:
In #4527 I retained the same `@CheckForNull` behavior since I could not come up with a better alternative. Returning a subdirectory of the workspace (as I think this patch does) is potentially unsafe: one expectation of a workspace temp dir is that it _not_ be made visible from `DirectoryBrowserSupport`.

Reply:
Okay, thank you. I will look at that one a little better